com.unity.perception/Runtime/Randomization/Randomizers/AssetSources/AssetSource.cs

```diff
@@ -0,0 +1,139 @@
+
+        /// <summary>
+        /// Returns the unprocessed asset loaded from the provided index
+        /// </summary>
+        /// <param name="index">The index of the asset to load</param>
+        /// <returns>The asset loaded at the provided index</returns>
+        public T LoadRawAsset(int index)
+        {
+            CheckIfInitialized();
+            return assetSourceLocation.LoadAsset<T>(index);
+        }
+
+        /// <summary>
+        /// Returns all unprocessed assets that can be loaded from this AssetSource
+        /// </summary>
+        /// <returns>All assets that can be loaded from this AssetSource</returns>
+        public T[] LoadAllRawAssets()
+        {
+            var array = new T[Count];
+            for (var i = 0; i < Count; i++)
+            {
+                array[i] = LoadRawAsset(i);
+                archetype.Preprocess(array[i]);
+            }
+            return array;
+        }
+
+        /// <summary>
+        /// Creates an instance of the asset loaded from the provided index and preprocesses it using the archetype
+        /// assigned to this asset source
+        /// </summary>
+        /// <param name="index">The index of the asset to load</param>
+        /// <returns>The instantiated instance</returns>
+        public T CreateProcessedInstance(int index)
+        {
+            CheckIfInitialized();
+            return CreateProcessedInstance(LoadRawAsset(index));
+        }
+
+        /// <summary>
+        /// Instantiates, preprocesses, and returns all assets that can be loaded from this asset source
+        /// </summary>
+        /// <returns>Instantiated instances from every loadable asset</returns>
+        public T[] CreateProcessedInstances()
+        {
+            var array = new T[Count];
+            for (var i = 0; i < Count; i++)
+                array[i] = CreateProcessedInstance(i);
+            return array;
+        }
+
+        /// <summary>
+        /// Returns a uniformly random sampled asset from this AssetSource
+        /// </summary>
+        /// <returns>The randomly sampled asset</returns>
+        public T SampleAsset()
+        {
+            CheckIfInitialized();
+            return assetSourceLocation.LoadAsset<T>((int)(m_Sampler.Sample() * Count));
+        }
+
+        /// <summary>
+        /// Instantiates and preprocesses a uniformly random sampled asset from this AssetSource
+        /// </summary>
+        /// <returns>The generated random instance</returns>
+        public T SampleInstance()
+        {
+            CheckIfInitialized();
+            return CreateProcessedInstance(SampleAsset());
+        }
+
+        /// <summary>
+        /// Unloads all assets that have been loaded from this AssetSource
+        /// </summary>
+        public void ReleaseAssets()
+        {
+            CheckIfInitialized();
+            assetSourceLocation.ReleaseAssets();
+        }
+
+        void CheckIfInitialized()
+        {
+            if (!m_Initialized)
```

Comment:
Could this CheckIfInitialized call be in the AssetSourceLocation itself?

Reply:
I did this originally, but I moved this to the AssetSource class so that people implementing asset sources wouldn't have to duplicate the manual process of checking if the asset source has been initialized before sampling assets.

I can move it back though.

Reply:
I've decided to leave this in the AssetSource